From 731959ef716bb72ba248616bca7241c8abe051e0 Mon Sep 17 00:00:00 2001
From: llin4 <ling.lin@intel.com>
Date: Wed, 13 Nov 2013 10:19:45 +0800
Subject: [PATCH 3/9] [PATCH] [PORT FROM MAIN] core: houdini hook rebase on KK

system/core houdini hook rebase on KK

Change-Id: Ibdb009940a272c407634fdffa0e63f68dc05fc58
Category: feature differentiation
Domain: AOSP.Houdini-Integration
Origin: internal
Signed-off-by: llin4 <ling.lin@intel.com>
---
 init/init.c        | 15 +++++++++++++++
 toolbox/Android.mk |  5 +++++
 toolbox/cat.c      | 12 ++++++++++++
 3 files changed, 32 insertions(+)

diff --git a/init/init.c b/init/init.c
index 53e0dae..a191c9a 100644
--- a/init/init.c
+++ b/init/init.c
@@ -31,6 +31,7 @@
 #include <sys/types.h>
 #include <sys/socket.h>
 #include <sys/un.h>
+#include <sys/personality.h>
 
 #include <selinux/selinux.h>
 #include <selinux/label.h>
@@ -916,6 +917,19 @@ static int property_service_init_action(int nargs, char **args)
     return 0;
 }
 
+static int personality_init_action(int nargs, char **args)
+{
+    char tmp[PROP_VALUE_MAX];
+    int ret;
+    ret = property_get("ro.config.personality", tmp);
+    if (ret && !strcmp(tmp, "compat_layout")) {
+        int old_personality;
+        old_personality = personality((unsigned long)-1);
+        personality(old_personality | ADDR_COMPAT_LAYOUT);
+    }
+    return 0;
+}
+
 static int signal_init_action(int nargs, char **args)
 {
     signal_init();
@@ -1229,6 +1243,7 @@ int main(int argc, char **argv)
     queue_builtin_action(mix_hwrng_into_linux_rng_action, "mix_hwrng_into_linux_rng");
 
     queue_builtin_action(property_service_init_action, "property_service_init");
+    queue_builtin_action(personality_init_action, "personality_init");
     queue_builtin_action(signal_init_action, "signal_init");
     queue_builtin_action(check_startup_action, "check_startup");
 
diff --git a/toolbox/Android.mk b/toolbox/Android.mk
index 3e28494..135ee26 100644
--- a/toolbox/Android.mk
+++ b/toolbox/Android.mk
@@ -99,6 +99,11 @@ LOCAL_SHARED_LIBRARIES := \
 	libusbhost \
 	libselinux
 
+ifeq ($(INTEL_HOUDINI), true)
+    LOCAL_CFLAGS += -DWITH_HOUDINI
+    LOCAL_STATIC_LIBRARIES += houdini_hook
+endif
+
 LOCAL_MODULE := toolbox
 
 # Including this will define $(intermediates).
diff --git a/toolbox/cat.c b/toolbox/cat.c
index 6ac31f8..c80cbd3 100644
--- a/toolbox/cat.c
+++ b/toolbox/cat.c
@@ -183,6 +183,10 @@ raw_cat(int rfd)
 	}
 }
 
+#ifdef WITH_HOUDINI
+extern int houdini_hook_open(const char *path, int flags, int mode);
+#endif
+
 static void
 raw_args(char **argv)
 {
@@ -196,7 +200,11 @@ raw_args(char **argv)
 				fd = fileno(stdin);
 			else if (fflag) {
 				struct stat st;
+#ifdef WITH_HOUDINI
+				fd = houdini_hook_open(*argv, O_RDONLY|O_NONBLOCK, 0);
+#else
 				fd = open(*argv, O_RDONLY|O_NONBLOCK, 0);
+#endif
 				if (fd < 0)
 					goto skip;
 
@@ -210,7 +218,11 @@ raw_args(char **argv)
 					goto skipnomsg;
 				}
 			}
+#ifdef WITH_HOUDINI
+			else if ((fd = houdini_hook_open(*argv, O_RDONLY, 0)) < 0) {
+#else
 			else if ((fd = open(*argv, O_RDONLY, 0)) < 0) {
+#endif
 skip:
 				perror(*argv);
 skipnomsg:
-- 
2.7.4

